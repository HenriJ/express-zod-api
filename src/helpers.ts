import * as z from 'zod';
import {MiddlewareDefinition} from './middleware';

export type ObjectSchema<T extends z.ZodRawShape> = z.ZodObject<T, "passthrough" | "strict" | "strip">;
export type Unshape<T> = T extends z.ZodRawShape ? z.infer<ObjectSchema<T>> : T;
export type JoinUnshaped<A, B> = Unshape<A> & Unshape<B>;

export function combineEndpointAndMiddlewareInputSchemas<IN extends z.ZodRawShape, mIN>(
  input: ObjectSchema<IN>,
  middlewares: MiddlewareDefinition<any, any, any>[]
): ObjectSchema<IN & mIN> {
  if (middlewares.length === 0) {
    return input as any as ObjectSchema<IN & mIN>;
  }
  return middlewares
    .map((middleware) => middleware.input)
    .reduce((carry, schema) => carry.merge(schema))
    .merge(input) as ObjectSchema<IN & mIN>;
}
