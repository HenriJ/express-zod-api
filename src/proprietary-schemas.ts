import { z } from "zod";
import { ZodDateIn } from "./date-in-schema";
import { ZodDateOut } from "./date-out-schema";
import { ZodFile } from "./file-schema";
import { getUploadSchema } from "./upload-schema";

export const file = ZodFile.create;
export const upload = getUploadSchema;
export const dateIn = ZodDateIn.create;
export const dateOut = ZodDateOut.create;

/** Shorthand for z.object({ raw: z.file().buffer() }) */
export const raw = () => z.object({ raw: ZodFile.create().buffer() });
