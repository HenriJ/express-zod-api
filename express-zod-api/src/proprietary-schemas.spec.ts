import { serializeSchema } from "../helpers/serializer";
import { raw } from "./proprietary-schemas";

describe("Proprietary schemas", () => {
  describe("ez.raw()", () => {
    test("Should an alias to ZodObject with raw:file", () => {
      expect(serializeSchema(raw())).toMatchSnapshot();
    });
  });
});
