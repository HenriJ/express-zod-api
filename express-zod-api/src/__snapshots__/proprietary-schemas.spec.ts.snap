// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proprietary schemas ez.raw() Should an alias to ZodObject with raw:file 1`] = `
{
  "_type": "ZodObject",
  "shape": {
    "raw": {
      "_type": "ZodFile",
      "encoding": undefined,
      "type": "buffer",
    },
  },
}
`;
